numpy/testing/_private/utils.py

```diff
@@ -769,16 +769,19 @@ def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
         y_id = func(y)
         # We include work-arounds here to handle three types of slightly
         # pathological ndarray subclasses:
-        # (1) all() on `masked` array scalars can return masked arrays, so we
-        #     use != True
+        # (1) all() on fully masked arrays returns np.ma.masked, so we use != True
+        #     (np.ma.masked != True evaluates as np.ma.masked, which is falsy).
         # (2) __eq__ on some ndarray subclasses returns Python booleans
-        #     instead of element-wise comparisons, so we cast to np.bool() and
-        #     use isinstance(..., bool) checks
+        #     instead of element-wise comparisons, so we cast to np.bool in
+        #     that case.
         # (3) subclasses with bare-bones __array_function__ implementations may
         #     not implement np.all(), so favor using the .all() method
-        # We are not committed to supporting such subclasses, but it's nice to
+        # We are not committed to supporting cases (2) and (3), but it's nice to
         # support them if possible.
-        if np.bool(x_id == y_id).all() != True:
+        result = x_id == y_id
+        if isinstance(result, bool):
+            result = np.bool(result)
+        if result.all() != True:
```

Comment:
Playing devil's advocate: what if someone passes something into this routine that returns something that's not a bool from `==` but that also doesn't have an `all()` method. You've introduced a new `AttributeError` that the cast to `np.bool` used to guard against.

Reply:
Should be handled now

Reply:
Thanks, this is still ugly but is better :)

Unfortunately with masked arrays there's an irreducible level of ugliness.